Content.Server/Radio/EntitySystems/JammerSystem.cs

```diff
@@ -0,0 +1,81 @@
+using Content.Shared.PowerCell.Components;
+
+namespace Content.Server.Radio.EntitySystems;
+
+public sealed class JammerSystem : EntitySystem
+{
+    [Dependency] private readonly PowerCellSystem _powerCell = default!;
+    [Dependency] private readonly PopupSystem _popup = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<RadioJammerComponent, ActivateInWorldEvent>(OnActivate);
+        SubscribeLocalEvent<RadioJammerComponent, PowerCellChangedEvent>(OnPowerCellChanged);
+        SubscribeLocalEvent<RadioJammerComponent, ExaminedEvent>(OnExamine);
+        SubscribeLocalEvent<RadioReceiveAttemptEvent>(OnRadioSendAttempt);
+    }
+
+    public override void Update(float frameTime)
+    {
+        foreach (var jam in EntityQuery<RadioJammerComponent>())
+        {
+            var uid = jam.Owner;
+            if (jam.Activated && _powerCell.TryGetBatteryFromSlot(uid, out var battery))
+                jam.Activated = battery.TryUseCharge(jam.Wattage * frameTime);
+        }
+    }
+
+    private void OnActivate(EntityUid uid, RadioJammerComponent comp, ActivateInWorldEvent args)
+    {
+        comp.Activated = 
+            !comp.Activated &&
+            _powerCell.TryGetBatteryFromSlot(uid, out var battery) &&
+            battery.CurrentCharge > comp.Wattage;
+        var state = Loc.GetString(comp.Activated ? "radio-jammer-component-on-state" : "radio-jammer-component-off-state");
+        var message = Loc.GetString("radio-jammer-component-on-use", ("state", state));
+        _popup.PopupEntity(message, args.User, args.User);
+        args.Handled = true;
+    }
+
+    private void OnPowerCellChanged(EntityUid uid, RadioJammerComponent comp, PowerCellChangedEvent args)
+    {
+        if (args.Ejected)
+            comp.Activated = false;
+    }
+
+    private void OnExamine(EntityUid uid, RadioJammerComponent comp, ExaminedEvent args)
+    {
+        if (args.IsInDetailsRange)
+        {
+            var msg = comp.Activated
+                ? Loc.GetString("radio-jammer-component-examine-on-state")
+                : Loc.GetString("radio-jammer-component-examine-off-state");
+            args.PushMarkup(msg);
+            if (_powerCell.TryGetBatteryFromSlot(uid, out var battery))
+                args.PushMarkup(Loc.GetString("radio-jammer-component-charge",
+                    ("charge", (int) ((battery.CurrentCharge / battery.MaxCharge) * 100))));
+        }
+    }
+
+    private void OnRadioSendAttempt(ref RadioReceiveAttemptEvent args)
+    {
+        if (args.RadioSource == null)
+            return;
+        var source = Transform(args.RadioSource.Value).Coordinates;
+        foreach (var jam in EntityQuery<RadioJammerComponent>())
+        {
+            if (jam.Activated && source.InRange(EntityManager, Transform(jam.Owner).Coordinates, jam.Range))
+            {
+                args.Cancelled = true;
+                return;
+            }
+        }
```

Comment:
Use EntityQuery<TransformComponent> and also check MapId first as it will make the check a lot faster.